### PR TITLE
[utils] Slab memory allocator

### PR DESCRIPTION
**Pull request by arnikola**
Allows strongly bounded slab allocations of memory in highly concurrent instances

### CODE REVIEW

File: src/dbnode/x/slab/types.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	FixedBufferCapacity() int
+
+	// SetFixedBufferTimeout sets the fixed buffer timeout, 0 implies no timeout.
+	SetFixedBufferTimeout(value time.Duration) Options
+
+	// FixedBufferTimeout returns the fixed buffer timeout, 0 implies no timeout.
+	FixedBufferTimeout() time.Duration
+
+	// Block()
+}
```
Comment (vdarulis) [inline src/dbnode/x/slab/types.go:71]: can we please not use the legacy options interface? see https://github.com/uber-go/guide/blob/master/style.md#functional-options

**Comment by mway** [comment]
a few questions:

1. would you mind moving this to `src/x/dbnode/...` instead of `src/dbnode/x/...`?
2. why "slab"?
3. also, does this need its own package? could we define these types in src/x/dbnode without introducing another package? or is there other stuff that will go in here?

File: src/dbnode/x/slab/types.go (hunk 1 above)
Comment (arnikola) [reply src/dbnode/x/slab/types.go]: Honestly the proposed options there seem much clunkier to use than this style; I agree that it's better than requiring long arg lists as in the bad column there, but not certain that the proposed pattern is any easier or more flexible than this approach, especially for building options that are not expected to change much

Comment (mway) [reply src/dbnode/x/slab/types.go]: There are a lot of reasons that we should be preferring the other options, but one major one is that we should not be building stateful options. Options should have no opinion on ordering, use each other, do "work", etc.

Using this "options interface type" approach means that now we need to update mocks if they exist, understand if there's some special behavior in one place that's different than a similar version, there's a bunch of extra type definition, etc. I'd flip the argument on its head: I don't see any substantive value in maintaining the current options pattern versus switching to a more conventional one.

Obviously it's something that pervades a lot of the codebase, but it's something that we should be moving away from.

File: src/x/buffer/fixed_buffer_prop_test.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+			fmt.Println("start", startStr)
+			fmt.Println("end  ", endStr)
+			return true, nil
+		},
+		genLongBytesChunked(numberOfBuffers, bufferCapacity),
+		genTakeByteSlices(numberOfBuffers, bufferCapacity),
+	))
+
+	// genParam := parameters.(gen.GenParameters)
+	// fmt.Println(genParam.NextUInt8())
```
Comment (robskillington) [inline src/x/buffer/fixed_buffer_prop_test.go:146]: Can probably remove this?

**Pull request closed without merging**